openlr/paths_connector.cpp

```diff
@@ -0,0 +1,220 @@
+  // TODO(mgsergio): Do not iterate more than kMaxRetries times.
+  // TODO(mgserigio): Make kMaxRetries depend on points number in the segment.
+  for (size_t i = 1; i < points.size(); ++i)
+  {
+    bool found = false;
+
+    auto const distanceToNextPoint = points[i - 1].m_distanceToNextPoint;
+    auto const & fromCandidates = lineCandidates[i - 1];
+    auto const & toCandidates = lineCandidates[i];
+
+    Graph::EdgeVector fakePath;
+
+    for (size_t fromInd = 0; fromInd < fromCandidates.size() && !found; ++fromInd)
+    {
+      for (size_t toInd = 0; toInd < toCandidates.size() && !found; ++toInd)
+      {
+        resultPath[i - 1].clear();
+
+        found = ConnectAdjacentCandidateLines(points[i - 1].m_functionalRoadClass,
+                                              distanceToNextPoint, fromCandidates[fromInd],
+                                              toCandidates[toInd], resultPath[i - 1]);
+
+        if (!found)
+          continue;
+
+        found = ValidatePath(resultPath[i - 1], distanceToNextPoint, m_pathLengthTolerance);
+        if (fakePath.empty() && found &&
+            (resultPath[i - 1].front().IsFake() || resultPath[i - 1].back().IsFake()))
+        {
+          fakePath = resultPath[i - 1];
+          found = false;
+        }
+      }
+    }
+
+    if (!fakePath.empty() && !found)
+    {
+      found = true;
+      resultPath[i - 1] = fakePath;
+    }
+
+    if (!found)
+    {
+      LOG(LDEBUG, ("No shortest path found"));
+      ++m_stat.m_noShortestPathFound;
+      resultPath[i - 1].clear();
+      return false;
+    }
+  }
+
+  ASSERT_EQUAL(resultPath.size(), points.size() - 1, ());
+
+  return true;
+}
+
+bool PathsConnector::FindShortestPath(FunctionalRoadClass const frc, uint32_t const maxPathLength,
+                                      Graph::Edge const & from, Graph::Edge const & to,
+                                      Graph::EdgeVector & path)
+{
+  // TODO(mgsergio): Turn dijkstra to A*.
+
+  ScopedTimer t(m_stat.m_shortestPathsTime);
+
+  uint32_t const kLengthToleranceM = 10;
+
+  struct State
+  {
+    State(Graph::Edge const & e, uint32_t const s) : m_edge(e), m_score(s) {}
+
+    bool operator>(State const & o) const
+    {
+      return make_tuple(m_score, m_edge) > make_tuple(o.m_score, o.m_edge);
+    }
+
+    Graph::Edge m_edge;
+    uint32_t m_score;
+  };
+
+  priority_queue<State, vector<State>, greater<State>> q;
+  map<Graph::Edge, uint32_t> scores;
+  map<Graph::Edge, Graph::Edge> links;
+
+  q.emplace(from, 0);
+  scores[from] = 0;
+
+  while (!q.empty())
+  {
+    auto const state = q.top();
+    q.pop();
+
+    auto const & u = state.m_edge;
+    // TODO(mgsergio): Unify names: use either score or distance.
+    auto const uScore = state.m_score;
+
+    if (uScore > maxPathLength + kLengthToleranceM)
+      continue;
+
+    if (uScore > scores[u])
+      continue;
+
+    if (u == to)
+    {
+      for (auto e = u; e != from; e = links[e])
+        path.push_back(e);
```

Comment:
How about to push ```from``` too?

Reply:
Why?

Reply:
Because the path starts from ```from```, right?

Reply:
Well have x1,x2...,from, from ... in the result path if I do as you ask. So non need to add twice.